### ISSUE TITLE
MTL for multiple modules?

### ISSUE THREAD

**Issue by pcheng2** [AvivNavon/nash-mtl#10]
Thanks for your owsome works, it's really cooooool and helpful! 

My question is about deploying the MTL methods on multiple NN modules. For example,

```python
import torch.nn.functional as F
import torch

class First_model(nn.Module):
    def __init__(self, 
                 input_dim, 
                 out_dim,
                 ):
        super(First_model, self).__init__()
        
        self.nn = nn.Sequential(
            torch.nn.Linear(self.input_dim, 128),
            torch.nn.ReLU(),
            torch.nn.Linear(128, 128),
            torch.nn.ReLU(),
            torch.nn.Linear(128, self.state_dim)
        )
    
    def forward(self, inputs):
        return self.nn(inputs)

class Second_model(nn.Module):
    def __init__(self, 
                 input_dim, 
                 out_dim,
                 ):
        super(Second_model, self).__init__()
        
        self.nn = nn.Sequential(
            torch.nn.Linear(self.input_dim, 128),
            torch.nn.ReLU(),
            torch.nn.Linear(128, 128),
            torch.nn.ReLU(),
            torch.nn.Linear(128, self.state_dim)
        )
    
    def forward(self, inputs):
        return self.nn(inputs)

class Third_model(nn.Module):
    def __init__(self, 
                 input_dim, 
                 out_dim,
                 ):
        super(Third_model, self).__init__()
        
        self.nn = nn.Sequential(
            torch.nn.Linear(self.input_dim, 128),
            torch.nn.ReLU(),
            torch.nn.Linear(128, 128),
            torch.nn.ReLU(),
            torch.nn.Linear(128, self.state_dim)
        )
    
    def forward(self, inputs):
        return self.nn(inputs)

first_model = First_model(25, 1)
second_model = Second_model(25, 1)
third_model = Third_model(25, 1)

x = torch.randn(10,25)
y = torch.randn(10,1)
m = torch.randn(10,1)
n = torch.randn(10,1)

losses.append(F.mse(first_model(x), y)) 
losses.append(F.mse(second_model(x), m)) 
losses.append(F.mse(third_model(x), n)) 
losses.append(F.mse(third_model(x), y)) 

``` 
Now I want to use MTL methods to train these models, but I don't know if I was using it correctly. Here is my template code:

```python

weight_method = WeightMethods(
        method,
        n_tasks=4, 
        device=device,
        **weight_method_params[method],
    )
loss, extra_outputs = weight_method.backward(
                losses=losses,
                shared_parameters=,
                task_specific_parameters=,
                last_shared_parameters=,
                representation=features,
            )
```
I'm not quite certain about what kind of variable structure should I feed in to  'weight_method.backward' in such case?

**Comment by AvivNavon** [comment]
Hi @pcheng2 
If I understand correctly the code you provided, you do not share parameters between the different tasks. For example, the `second_model` is only used to predict the "m" task, and the `third_model` is only used to predict the "n" task. In that case all parameters are tasks specific and so no MTL method is needed (you are basically doing single task learning in parallel for several tasks). If the tasks are related in any way, consider using a shared module (`shared_model`) to map `x` into features `z` which will than be feed into the task specific modules (`first_model`, `second_model`, ...). In that case the parameters of the `shared_model` are the `shared_parameters` and all other params are the `task_specific_parameters`.
Hope that helps

**Comment by pcheng2** [comment]
Hi @AvivNavon,

Thank you so much for your replying! I'm sorry for the confusion about the question because I'm literally not quite familiar with MTL and based on your patient explanation, I'm more confident that I was misunderstanding the problem settings in MTL.  

In my practical work, I will get different loss terms (approx. 8) from various modules(encoder-decoder and some feed-foward models), and backpropagate these losses to the modules' parameters. The problem is that I found these loss terms have different scales, thus the whole module may overfit in some direction where other losses are not converged yet.  And my current solution is to manullay assign different weights on these terms.

Therefore, I was wrongly thought I might encounter an MTL problem, and try to find a method to automatically allocate the weights of these loss terms and their gradients. While in my case, each model is doing 'single task learning', there is no parameter/module sharing between them. I'll try your advice about using a shared module with different task-specific modules. 

And for MTL tasks, I'm wondering that is it a necessary condition for the model to have various heads (usually at the last layer and with identical logits as input) regard to the task number? 

Anyway, thank you so much for your kind response and suggestion, I'll try it out!
